### PR TITLE
Add `static_map::retrieve_all`

### PR DESCRIPTION
**Pull request by PointKernel**
This PR adds a new API: `static_map::retrieve_all` which retrieves all entries in `static_map`. It could be used to help `cudf::groupby::hash` refactoring.

### CODE REVIEW

**Comment by PointKernel** [comment]
rerun tests

**Comment by PointKernel** [comment]
rerun tests

**Comment by sevagh** [comment]
rerun tests

**Comment by sevagh** [comment]
rerun tests

File: tests/static_map/duplicate_keys_test.cu [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+  {
+    map.insert(pairs_begin, pairs_begin + num_keys);
+
+    auto const num_entries = map.get_size();
+    REQUIRE(num_entries == num_keys / 2);
+
+    thrust::device_vector<Key> unique_keys(num_entries);
+    auto values_begin = thrust::make_discard_iterator();
+
+    map.retrieve_all(unique_keys.begin(), values_begin);
```
Comment (jrhemstad) [inline tests/static_map/duplicate_keys_test.cu]: Seems like we should also test `retrieve_all` from an empty map.

File: include/cuco/static_map.cuh [hunk 2]
```diff
@@ -296,6 +296,26 @@ class static_map {
...
+   * @tparam KeyOut Device accessible random access output iterator whose `value_type` is
+   * convertible from `key_type`.
+   * @tparam ValueOut Device accesible random access output iterator whose `value_type` is
+   * convertible from `mapped_type`.
+   * @param keys_out Beginning output iterator for keys
+   * @param values_out Beginning output iterator for values
+   * @param stream CUDA stream used for this operation
+   */
+  template <typename KeyOut, typename ValueOut>
+  void retrieve_all(KeyOut keys_out, ValueOut values_out, cudaStream_t stream = 0);
```
Comment (jrhemstad) [inline include/cuco/static_map.cuh]: This should either return the number of retrieved keys or return a `pair<KeyOut, ValueOut>` iterators to the end of the two output ranges.

File: include/cuco/static_map.cuh [hunk 3]
```diff
@@ -296,6 +296,26 @@ class static_map {
...
             cudaStream_t stream = 0);
 
+  /**
+   * @brief Retrieves all of the keys and their associated values.
+   *
+   * The order in which keys are returned is implementation defined and not guaranteed to be
+   * consistent between subsequent calls to `retrieve_all`.
+   *
+   * Behavior is undefined if the range beginning at `keys_out` or `values_out` is not large enough
+   * to contain the number of keys in the map.
```
Comment (jrhemstad) [inline include/cuco/static_map.cuh]:
```suggestion
   *
   * Behavior is undefined if the range beginning at `keys_out` or `values_out` is less than `size()`
```

**Comment by sevagh** [comment]
rerun tests

File: tests/static_map/duplicate_keys_test.cu (hunk 1 above)
Comment (PointKernel) [reply tests/static_map/duplicate_keys_test.cu]: Good point.

File: include/cuco/static_map.cuh (hunk 3 above)
Comment (PointKernel) [reply include/cuco/static_map.cuh]: Are you referring to `get_size()`? Time to remove all `get_` prefixes.

Comment (jrhemstad) [reply include/cuco/static_map.cuh]: ugh, yeah.

File: include/cuco/static_map.cuh (hunk 2 above)
Comment (PointKernel) [reply include/cuco/static_map.cuh]: Update API to return a pair of iterators pointing to the end of the output.

File: tests/static_map/duplicate_keys_test.cu [hunk 4]
```diff
@@ -63,9 +63,10 @@ TEMPLATE_TEST_CASE_SIG("Duplicate keys",
     thrust::device_vector<Key> unique_keys(num_entries);
     auto values_begin = thrust::make_discard_iterator();
 
-    map.retrieve_all(unique_keys.begin(), values_begin);
-    thrust::sort(thrust::device, unique_keys.begin(), unique_keys.end());
+    auto [key_out_end, value_out_end] = map.retrieve_all(unique_keys.begin(), values_begin);
+    REQUIRE(std::distance(unique_keys.begin(), key_out_end) == map.get_size());
```
Comment (jrhemstad) [inline tests/static_map/duplicate_keys_test.cu]: Test the `value_out_end` too?

**Review by jrhemstad — APPROVED** [review]
